### PR TITLE
feat: implement `CairoLayout`

### LINKED ISSUE #228: feat: implement Cairo layout
## Background

We're currently in the process of implementing various builtins within our project. Alongside this crucial development phase, we recognize the need to establish a robust architecture that facilitates the definition and integration of these builtins into the Cairo layout.

## Objective

The goal of this task is to devise a comprehensive architecture that accommodates the incorporation of different builtins seamlessly into the Cairo layout. This architectural blueprint will serve as the foundational framework for integrating and organizing the diverse set of builtins being implemented concurrently.

### PR DESCRIPTION
**Pull request by tcoratger**
Should close #228.

### CODE REVIEW

**Comment by tcoratger** [comment]
@lana-shanghai As this is `CairoLayout` and it is possible, in unit tests, to check each parameter in a hardcoded way, I did it. This is obviously a lot of lines of code but this way we are 100% sure and we will always see any changes. Don't hesitate to tell me if it's too many tests I can refactor, but I said to myself that I might as well test everything.

File: src/lib.zig [hunk 1]
```diff
@@ -13,6 +13,9 @@ pub const vm = struct {
     pub usingnamespace @import("vm/runners/cairo_runner.zig");
     pub usingnamespace @import("vm/builtins/bitwise/bitwise.zig");
     pub usingnamespace @import("vm/types/builtins_instance_def.zig");
+    pub usingnamespace @import("vm/types/diluted_pool_instance_def.zig");
+    pub usingnamespace @import("vm/types/cpu_instance_def.zig");
+    pub usingnamespace @import("vm/types/layout.zig");
```
Comment (lana-shanghai) [inline src/lib.zig:18]: Let's make sure we finish implementing the builtins before we touch the layout topic

File: src/vm/types/layout.zig [hunk 2]
```diff
@@ -1,8 +1,946 @@
...
+    /// defining specific parameters for CPU, Range Check units, memory allocation, and trace columns.
+    ///
+    /// Default parameters:
+    /// - `cpu_component_step`: 1
+    /// - `rc_units`: 4
+    /// - `public_memory_fraction`: 4
+    /// - `memory_units_per_step`: 8
+    /// - `n_trace_columns`: 22
+    /// - Other instances set to default or null.
+    pub fn dexInstance() Self {
```
Comment (lana-shanghai) [inline src/vm/types/layout.zig:108]: I think these instance definitions are redundant at the moment, let's discuss the ones we need after the builtins for the milestone. We also have to establish a commit which marks the end of the first milestone.

**Review by lana-shanghai — CHANGES_REQUESTED** [review]

**Comment by tcoratger** [comment]
@lana-shanghai FYI I just removed the unnecessary layouts leaving only the following layouts:
- Plain
- Small
- All Cairo
- Dynamic (I added it in addition to those that we mentioned yesterday because it is a configurable layout which could be useful very soon in my opinion)

**Review by lana-shanghai — APPROVED** [review]
lgtm

**Pull request merged**